File: sonar-testing-harness/src/main/java/org/sonar/test/TestUtils.java
/*
 * SonarQube, open source software quality management tool.
 * Copyright (C) 2008-2014 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * SonarQube is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * SonarQube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.test;

import org.apache.commons.io.FileUtils;
import org.apache.commons.lang.StringUtils;

import java.io.File;
import java.lang.reflect.Constructor;
import java.lang.reflect.Modifier;
import java.net.URL;

/**
 * Utilities for unit tests
 *
 * @since 2.2
 */
public final class TestUtils {

  private TestUtils() {
  }

  /**
   * Search for a test resource in the classpath. For example getResource("org/sonar/MyClass/foo.txt");
   *
   * @param path the starting slash is optional
   * @return the resource. Null if resource not found
   */
  public static File getResource(String path) {
    String resourcePath = path;
    if (!resourcePath.startsWith("/")) {
      resourcePath = "/" + resourcePath;
    }
    URL url = TestUtils.class.getResource(resourcePath);
    if (url != null) {
      return FileUtils.toFile(url);
    }
    return null;
  }

  /**
   * Search for a resource in the classpath. For example calling the method getResource(getClass(), "myTestName/foo.txt") from
   * the class org.sonar.Foo loads the file $basedir/src/test/resources/org/sonar/Foo/myTestName/foo.txt
   *
   * @return the resource. Null if resource not found
   */
  public static File getResource(Class baseClass, String path) {
    String resourcePath = StringUtils.replaceChars(baseClass.getCanonicalName(), '.', '/');
    if (!path.startsWith("/")) {
      resourcePath += "/";
    }
    resourcePath += path;
    return getResource(resourcePath);
  }

  /**
   * Asserts that all constructors are private, usually for helper classes with
   * only static methods. If a constructor does not have any parameters, then
   * it's instantiated.
   */
  public static boolean hasOnlyPrivateConstructors(Class clazz) {
    boolean ok = true;
    for (Constructor constructor : clazz.getDeclaredConstructors()) {
      ok &= Modifier.isPrivate(constructor.getModifiers());
      if (constructor.getParameterTypes().length == 0) {
        constructor.setAccessible(true);
        try {
          constructor.newInstance();
        } catch (Exception e) {
          throw new IllegalStateException(String.format("Fail to instantiate %s", clazz), e);
        }
      }
    }
    return ok;
  }

  public static File newTempDir(String prefix) {
    try {
      // Technique to create a temp directory from a temp file
      File f = File.createTempFile(prefix, "");
      f.delete();
      f.mkdir();
      return f;
    } catch (Exception e) {
      throw new IllegalStateException("Fail to create temp dir", e);
    }
  }
}

File: server/sonar-web/src/main/js/apps/overview/branches/test-utils.ts
/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import userEvent from '@testing-library/user-event';
import { byLabelText, byRole, byTestId, byText } from '~sonar-aligned/helpers/testSelector';
import { SoftwareImpactSeverity, SoftwareQuality } from '../../../types/clean-code-taxonomy';

export const getPageObjects = () => {
  const user = userEvent.setup();
  const selectors = {
    unsolvedOverallMessage: byText(/overview.ai_assurance.unsolved_overall.title/),
    dismissUnsolvedButton: byRole('button', {
      name: 'overview.ai_assurance.unsolved_overall.dismiss',
    }),
    overallCodeButton: byRole('tab', { name: /overview.overall_code/ }),
    softwareImpactMeasureCard: (softwareQuality: SoftwareQuality) =>
      byTestId(`overview__software-impact-card-${softwareQuality}`),
    softwareImpactMeasureCardRating: (softwareQuality: SoftwareQuality, rating: string) =>
      byLabelText(
        `overview.project.software_impact.has_rating.software_quality.${softwareQuality}.${rating}`,
      ),
  };
  const ui = {
    ...selectors,
    expectSoftwareImpactMeasureCard: (
      softwareQuality: SoftwareQuality,
      rating?: string,
      total?: number,
      branch = 'master',
      failed = false,
    ) => {
      if (failed) {
        expect(
          byTestId(`overview__software-impact-card-${softwareQuality}`)
            .byText('overview.measures.failed_badge')
            .get(),
        ).toBeInTheDocument();
      }

      if (typeof rating === 'string') {
        expect(
          byText(rating, { exact: true }).get(ui.softwareImpactMeasureCard(softwareQuality).get()),
        ).toBeInTheDocument();
      }
      if (total !== undefined) {
        const branchQuery = branch ? `&branch=${branch}` : '';

        expect(
          byRole('link', {
            name: `overview.measures.software_impact.see_list_of_x_open_issues.${total}.software_quality.${softwareQuality}`,
          }).get(),
        ).toHaveAttribute(
          'href',
          `/project/issues?issueStatuses=OPEN%2CCONFIRMED&impactSoftwareQualities=${softwareQuality}${branchQuery}&id=foo`,
        );
      }
    },
    expectSoftwareImpactMeasureCardToHaveOldMeasures: (
      softwareQuality: SoftwareQuality,
      rating: string,
      total: number,
      oldMetric: string,
      branch = 'master',
    ) => {
      const branchQuery = branch ? `&branch=${branch}` : '';
      expect(
        byText(rating, { exact: true }).get(ui.softwareImpactMeasureCard(softwareQuality).get()),
      ).toBeInTheDocument();
      expect(
        byRole('link', {
          name: `overview.measures.software_impact.see_list_of_x_open_issues.${total}.software_quality.${softwareQuality}`,
        }).get(),
      ).toHaveAttribute(
        'href',
        `/project/issues?issueStatuses=OPEN%2CCONFIRMED&types=${oldMetric}${branchQuery}&id=foo`,
      );
    },
    expectSoftwareImpactMeasureBreakdownCard: (
      softwareQuality: SoftwareQuality,
      severity: SoftwareImpactSeverity,
      active: boolean,
    ) => {
      const link = byTestId(
        `overview__software-impact-${softwareQuality}-severity-${severity}`,
      ).get(ui.softwareImpactMeasureCard(softwareQuality).get());
      if (active) {
        expect(link).toHaveClass('active');
      } else {
        expect(link).not.toHaveClass('active');
      }
    },
    expectSoftwareImpactMeasureCardRatingTooltip: async (
      softwareQuality: SoftwareQuality,
      rating: string,
      text: string,
    ) => {
      await expect(
        ui.softwareImpactMeasureCardRating(softwareQuality, rating).get(),
      ).toHaveATooltipWithContent(text);
    },
  };
  return { user, ui };
};
